feat: Add Dataset implementation

yeewen1018 opened the pull request:
## Overview
This PR introduces the `Dataset` abstraction and a concrete `InMemoryDataset` implementation. 

1. **`Dataset` trait** 
    - A unified interface for both in-memory and iterable datasets.
    - Requires `Send + Sync` to safely share across multiple worker threads. 

2. **`InMemoryDataset`**
    - Internally stores data as `Arc<[Sample]>`. Alternatives considered: 
       - `Vec<Sample>`: No zero-copy sharing 
       - `Vec<&Sample>`: Lifetime complexity 
    - **Iteration modes**: 
       - `.iter()` -> static dispatch (the default) - zero-cost abstraction, fastest for tight loops. 
       - `#[cfg(feature="boxed-iter")] .iter_boxed()` -> boxed iterator for dynamic pipelines. 
   - **Benchmarked on 1 M samples:**  
       | Operation        | Static Dispatch | Boxed Dispatch | Overhead |
       | ---------------- | --------------- | -------------- | -------- |
       | Pure iteration   | 12 ms           | 14 ms          | +16 %    |
       | With tensor ops  | 320 ms          | 322 ms         | +0.6 %   |
   - In tensor heavy workloads, boxed overhead is <1% (negligible), but static dispatch avoids heap allocations entirely.  

## Testing
- **Unit tests cover**: iteration correctness, concurrent access, and integration with `Collator`. 

## Next Steps 

1. **`IterableDataset`**
     - Streaming with sync workers (-> async in future). 

2. **`FileReader/ShardReader`**
     - Unified parsing (JSON/Parquet/Safetensors/Other format -> `Sample`)
 
3. **`Transform + Sampler`**
     - `Transform`: tokenization, resizing, normalization, data augmentation
     - `Sampler`: Shuffling/stratification

4. **`DataLoader`**
     - Worker pool + prefetching

```mermaid
flowchart TD
  subgraph "File Readers"
    J[JSON Reader]
    P[Parquet Reader]
    S[Safetensors Reader]
    O[Other Formats]
  end

  J -->|implements| FR[FileReader]
  P -->|implements| FR
  S -->|implements| FR
  O -->|implements| FR

  ITDS[IterableDataset]
  IMDS[InMemoryDataset]

  FR --> ITDS
  FR --> IMDS

  subgraph "Data Pipeline"
    ITDS -->|Iterator over Sample| DL[DataLoader]
    IMDS -->|Iterator over Sample| DL

    DL --> SM[Sampler]
    SM --> XF[Transform]
    XF --> CL[Collator]
    CL --> MB[MiniBatch]
  end
```

yeewen1018 committed:
feat: add Dataset trait and make Sample struct Send+Sync

yeewen1018 committed:
feat(dataset): implement InMemoryDataset

yeewen1018 committed:
test(dataset): add unit tests for InMemoryDataset

yeewen1018 committed:
test(dataset): add pipeline integration tests

Add end-to-end validation for:
- InMemoryDataset sample loading/iteration
- Stack collation overflow
- Padding collation with variable-length samples
- Device transfer roundtrips (CPU<->GPU)

yeewen1018 committed:
feat(dataset): refine Dataset trait docs and change get() return type
- Update documentation on the `Dataset` trait for `get`/`len` to clarify the expected behaviour of InMemoryDataset and IterableDataset.
- Change `get` function from `Option<Result<&Sample>>` to `Result<Option<&Sample>>` for better error handling.
- Update all affected tests to match the new `get` function.

yeewen1018 committed:
feat(dataset): add opt-in boxed iterator via `boxed-iter` feature.

yeewen1018 committed:
bench: add InMemoryDataset iteration benchmarks for static vs boxed iteration.

nikg7 commented on crates/data_preparation/src/dataset.rs:
```diff
@@ -0,0 +1,312 @@
+///
+/// All implementations must be `Send + Sync` to allow for safe
+/// sharing across threads.
+pub trait Dataset: Send + Sync {
```
Can we add a definition  (at least) of IterableDataset as well ?

it'd be useful to see the difference between `Dataset` and `IterableDataset`

nikg7 commented on crates/data_preparation/src/dataset.rs:
```diff
@@ -0,0 +1,312 @@
+    /// Random-access lookup by index.
+    /// - In-memory datasets return `Ok(Some(&Sample))` or `Ok(None)` if out-of-bounds.
+    /// - Streaming datasets always return `Ok(None)`.
+    fn get(&self, index: usize) -> Result<Option<&Sample>>;
```
I'm not 100% sure if methods `get(index)` and `len()`  returning `Option<...>` make sense in the base class  

Let's look at your proposed definition of `IterableDataset`  to decide. Thanks!!

yeewen1018 replied:
I moved `get`/`len` into `InMemoryDataset` (since `IterableDataset::get/len` would always return `None`) and added `IterableDataset` in commit [92a68ec](https://github.com/oumi-ai/roumi/pull/22/commits/92a68ec3a471753c57fc1e22350c86edaa5209c4). We could alternatively keep `get/len` on the trait and have `IterableDataset` return a type‐error there—let me know if you’d prefer that. Thank you.

yeewen1018 committed:
refactor(dataset): remove `len()`, `get()` and `is_empty()` from Dataset trait

- Add inherent `get(&self, idx) -> Option<&Sample>`, `len(&self) -> usize`, and `is_empty(&self) -> bool` to `InMemoryDataset`
- Update `len()` to return `usize` instead of `Option(usize)`
- Update tests to return `usize` for `len()` assertions

yeewen1018 committed:
feat(dataset): Add IterableDataset for streaming support

- Add `DataSource` enum (to support various types of data sources)
- Implement IterableDataset with dummy per-source `stream_samples()` method
- Include basic test for `IterableDatset::iter()`

nikg7 commented on crates/data_preparation/Cargo.toml:
```diff
@@ -9,4 +9,15 @@ description.workspace = true
+anyhow = "1.0"
+
+[features]
+boxed-iter = []
```
what's the reason for making it an optional feature?  Can we keep it always enabled to keep things simple?

nikg7 commented on crates/data_preparation/benches/dataset_bench.rs:
```diff
@@ -0,0 +1,104 @@
+        let ds = make_dataset(size);
+        group.throughput(Throughput::Elements(size as u64));
+
+        group.bench_with_input(BenchmarkId::new("static+map", size), &size, |b, &_s| {
```
Replace `|b, &_s|` with more self-descriptive names

nikg7 commented on crates/data_preparation/benches/dataset_bench.rs:
```diff
@@ -0,0 +1,104 @@
+        group.bench_with_input(BenchmarkId::new("static+map", size), &size, |b, &_s| {
+            b.iter(|| {
+                ds.iter()
+                    .map(|r| {
```
Replace `r` with a more self-descriptive name. Also for `t` below

... same for all other occurrences of a single character cryptic variable's names in this PR

nikg7 commented on crates/data_preparation/src/sample.rs:
```diff
@@ -19,6 +19,32 @@ pub struct Sample {
+///    * All tensor operations that modify data require a mutable `Tensor` handle,
+///      therefore we cannot mutate through just a shared `&Sample` reference.
+///    * Immutable tensor reads are thread-safe (as per tch-rs).
+unsafe impl Send for Sample {}
```
sorry, why it's `unsafe` ?

yeewen1018 replied:
Okay I have:
1. Removed the `boxed-iter` feature ([3db3e59](https://github.com/oumi-ai/roumi/pull/22/commits/3db3e591158905d78d7cbd9425c26e2388b9f924)),
2. Fixed the benchmark variable names [(7e7b6bc)](https://github.com/oumi-ai/roumi/pull/22/commits/7e7b6bcab0375eb33215172a5b3dcfd2c19f1e41), and
3. Updated the `unsafe impl` documentation in `Sample` with references to `tch::Tensor`'s source code (https://docs.rs/tch/latest/src/tch/wrappers/tensor.rs.html).

The `unsafe` marker here documents our thread-safety verification - it is a guarantee that all components (`HashMap`, `String`, and `Tensor`) are validated as thread-safe, not an indication of danger here. 

Thank you for the comments.

yeewen1018 committed:
refactor: remove boxed-iter feature flag
- Make boxed iterator always available

yeewen1018 committed:
refactor(benches): Clarify benchmark variable names
- b -> bench (benchmark harness)
- r -> sample_result (iterator item)
- t -> processed_tensor (transformed tensor)
- ds -> dataset
- &_s -> &_dataset_size

yeewen1018 committed:
docs(sample): document Send/Sync safety with tch-rs references

yeewen1018 committed:
refactor(dataset)!: remove old Dataset API

- Drop the old Dataset trait and its InMemoryDataset/IterableDataset implementations,
which assumed data was already in Sample form (HashMap<String, tch::Tensor>).
- This paves the way for a generic Raw->Sample pipeline where users can load
raw records (text, JSON, images, etc.) and then transform them into Samples.
- Remove dataset_bench.rs, since it needs to be rewritten against the new API.

BREAKING CHANGE: the old Dataset API has been removed.

yeewen1018 committed:
chore: remove `dataset-bench` benchmark target from Cargo.toml

yeewen1018 committed:
feat(dataset): add InMemoryDataset<Raw> and introduce Transform trait

- Implement InMemoryDataset<Raw> with new(), get(), len(), with_transform()
- Hook it up to Dataset::iter() to apply the transform pipeline
- Add unit test for InMemoryDataset

yeewen1018 committed:
feat(dataset): add DataSource trait and IterableDataset<Raw>
- Implement `IterableDataset<Raw>` to wrap one or more data sources and apply transforms
- Add unit tests verifying `IterableDataset` iteration and transform integration

yeewen1018 commented:
Hi @nikg4, I replaced the old `dataset.rs` to introduce a new definition for the `Dataset` abstraction. This should now allow the dataset to take in raw, unprocessed data and transform it into `Sample`. The previous definition for `Dataset` assumes pre-packaged `Sample`s but real-world data rarely comes pre-packaged in that format. 

- The raw data can be loaded from files, websites, databases, etc., via a new `DataSource<Raw>` trait. The streaming logic for these various sources will be implemented later. 
- I also introduced a `Transform<Raw, Sample>` trait to support preprocessing steps such as resize, normalization, and ToTensor (to be implemented later), along with a `Compose` helper (also to come) for chaining these transformations.
- Sharding logic will be handled by a `Sampler` trait in a future PR.

I look forward to your review—thank you!

yeewen1018 commented:
@nikg4  I just opened a follow-up PR #23  (`data_sources` → `dataset-trait`) that builds on this one. PR #23 implements the first batch of `DataSource` readers (`TxtSource`, `JsonlSource`, etc.).

No rush on reviewing #23 —it is just for visibility since it depends on #22 . Thank you for your time!
